### PR TITLE
masync: add flag indicating that future is async

### PR DESCRIPTION
**Pull request by DamianDuy**
The documentation to these changes will be added in another PR since this one is already growing too big.

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)

### CODE REVIEW

**Comment by kswiecicki** [comment]
*[`tests/runtime_test/runtime_test.c` line 52 at r1](https://reviewable.io/reviews/pmem/miniasync/100#-N3J6YzZ5M_U_4LIHg6r:-N3J6Yz_83fovML9Q6st:buhty57) ([raw file](https://github.com/pmem/miniasync/blob/52e6e2810083bad2dd5f84f361484eb79941d74d/tests/runtime_test/runtime_test.c#L52)):*
> ```C
> 			str_len / 2, 0);
> 
> 	struct vdm_operation_future memmove_fut_symbol_s_to_b =
> ```

It's a memset future.



_Code quote:_
```C
memmove_fut_symbol_s_to_b
```

**Comment by kswiecicki** [comment]
*[`src/include/libminiasync/future.h` line 151 at r1](https://reviewable.io/reviews/pmem/miniasync/100#-N3J6s0-0SMAke53fOXi:-N3J6s009Rb8GTJuJkdJ:b-j0baqv) ([raw file](https://github.com/pmem/miniasync/blob/52e6e2810083bad2dd5f84f361484eb79941d74d/src/include/libminiasync/future.h#L151)):*
> ```objc
> #define FUTURE_VALID_FLAGS (FUTURE_IS_ASYNC)
> 
> #define SET_FUTURE_FLAG(_futurerep, _flag)\
> ```

Looking at our other macros, I believe that `FUTURE_SET_FLAG` would be more fitting.



_Code quote:_
```objc
SET_FUTURE_FLAG
```

File: tests/runtime_test/runtime_test.c [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#include <string.h>
+#include "util_dml.h"
+#include "core/os.h"
+#include "test_helpers.h"
+
+/*
+ * Test executing multiple futures with different data movers
+ */
+int
+test_runtime_wait_multiple(size_t str_len)
```
Comment (lplewa) [inline tests/runtime_test/runtime_test.c:17]: You need to add test witch custom features (not using vdm one),

Test the following cases: 
- Chain future with has async and non async futures.
- Future which changes its flag during execution.

**Review by lplewa — CHANGES_REQUESTED** [review]

**Review by pbalcer — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: 0 of 12 files reviewed, 11 unresolved discussions (waiting on @DamianDuy, @kswiecicki, @lplewa, and @pbalcer)

___
*[`src/include/libminiasync/future.h` line 141 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oBon02FTEVHSpNWKe:-N5oBon02FTEVHSpNWKf:b-79dgbo) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L141)):*
> ```objc
> typedef enum future_state (*future_task_fn)(struct future_context *context,
> 			struct future_notifier *notifier);
> typedef int (*has_flag_fn)(void *future);
> ```

future_has_flag_fn

(symbols, especially public symbols, should be prefixed with their module name)
___
*[`src/include/libminiasync/future.h` line 144 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oCKPxCSVzjYJIzO9R:-N5oCKPyCZs5T2rmRjhl:bj3psdx) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L144)):*
> ```objc
> 
> struct future {
> 	uint64_t flags;
> ```

I think we can remove the flags parameter, and then just have a function to check whether a future is something, e.g.,:

```
enum FUTURE_PROPERTY {
 FUTURE_PROPERTY_ASYNC,
 FUTURE_PROPERTY_XXX,
};

if (future->has_property(future, FUTURE_PROPERTY_ASYNC)) {
 //
}
```

By default, these can all be false, and futures can choose to implement their own has_property function. Saves us 8 bytes and makes this whole thing simpler.
___
*[`src/include/libminiasync/future.h` line 150 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oB97W1w0oXmPwa1Ky:-N5oB97W1w0oXmPwa1Kz:bxjmq8w) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L150)):*
> ```objc
> };
> 
> #define FUTURE_IS_ASYNC		(((uint64_t)1) << 0)
> ```

FUTURE_FLAG_ASYNC
___
*[`src/include/libminiasync/future.h` line 166 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oF98p9tyvwHT2Db8_:-N5oF98p9tyvwHT2Db8a:b-q25mvp) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L166)):*
> ```objc
> 	}
> 
> #define FUTURE_INIT(_futurep, _taskfn)\
> ```

add something like FUTURE_INIT_EXT that also accepts a flag/property checking function, and then implement the other two macros using that macro.
___
*[`src/include/libminiasync/future.h` line 292 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oBb160VUY5nbrvkvv:-N5oBb175sdNrKZ4HubL:b7w6n8) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L292)):*
> ```objc
>  * chained futures.
>  */
> static inline int future_async_flag(struct future *fut)
> ```

future_check_flag/future_has_flag, and this should probably take a flag as a parameter. (it only works right now because you only have one flag :P)
___
*[`src/include/libminiasync/future.h` line 365 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oBMIG3gUjSWsueyyj:-N5oBMIG3gUjSWsueyyk:b-ist9co) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L365)):*
> ```objc
> }
> 
> static inline int basic_has_flag(void *future)
> ```

future_has_flag_base
___
*[`src/include/libminiasync/future.h` line 373 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oF2fT14lTqbgFgTnP:-N5oF2fT14lTqbgFgTnQ:b1biy12) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L373)):*
> ```objc
> }
> 
> static inline int chained_has_flag(void *future)
> ```

future_chain_has_flag
___
*[`src/include/libminiasync/future.h` line 386 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oFWqS82wAbQ66Knkj:-N5oFWqS82wAbQ66Knkk:b-41tl60) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L386)):*
> ```objc
> 	size_t used_data = 0;
> 
> 	while (entry != NULL) {
> ```

This is too complicated to just copy/paste. Write some generic function.
___
*[`src/include/libminiasync/future.h` line 403 at r2](https://reviewable.io/reviews/pmem/miniasync/100#-N5oFSNfEtK5LDkB3HZm:-N5oFSNfEtK5LDkB3HZn:b-uv7i6k) ([raw file](https://github.com/pmem/miniasync/blob/b578b207fcc212835c678e187076f922536f3a52/src/include/libminiasync/future.h#L403)):*
> ```objc
> 				if ((entry->future.flags & FUTURE_IS_ASYNC) != FUTURE_IS_ASYNC)
> 					return 0;
> 				return 1;	
> ```

remove whitespace

**Review by pbalcer — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: 0 of 12 files reviewed, 5 unresolved discussions (waiting on @DamianDuy, @kswiecicki, @lplewa, and @pbalcer)

___
*[`src/include/libminiasync/future.h` line 291 at r4](https://reviewable.io/reviews/pmem/miniasync/100#-N7KgoFX1BMHNZFXt5p1:-N7KgoFX1BMHNZFXt5p2:b-4rwh8v) ([raw file](https://github.com/pmem/miniasync/blob/5daae88e9d95e68a77def58acb0fec8e4d5445b0/src/include/libminiasync/future.h#L291)):*
> ```objc
>  * chained futures.
>  */
> static inline int fut_has_property(struct future *fut,
> ```

future_has_property

**Review by DamianDuy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: 0 of 12 files reviewed, 5 unresolved discussions (waiting on @kswiecicki, @lplewa, and @pbalcer)

___
*[`src/include/libminiasync/future.h` line 291 at r4](https://reviewable.io/reviews/pmem/miniasync/100#-N7KgoFX1BMHNZFXt5p1:-N7L1ih6-2dE_kOfGkDu:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/5daae88e9d95e68a77def58acb0fec8e4d5445b0/src/include/libminiasync/future.h#L291)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

future_has_property
</blockquote></details>

Done.

**Review by pbalcer — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: 0 of 12 files reviewed, 6 unresolved discussions (waiting on @DamianDuy, @kswiecicki, @lplewa, and @pbalcer)

___
*[`src/runtime.c` line 119 at r4](https://reviewable.io/reviews/pmem/miniasync/100#-N7LT_zZ1PXJfIro3pwS:-N7LT_zZ1PXJfIro3pwT:b-mp6mlh) ([raw file](https://github.com/pmem/miniasync/blob/5daae88e9d95e68a77def58acb0fec8e4d5445b0/src/runtime.c#L119)):*
> ```C
> 	for (;;) {
> 		for (uint64_t i = 0; i < runtime->spins_before_sleep; ++i) {
> 			quicksort_futs_async_property(futs, 0, (int)nfuts - 1);
> ```

use quick sort from libc (sorry, my fault - I should have been more clear) ;)
https://man7.org/linux/man-pages/man3/qsort.3.html

**Review by blazej-smorawski — COMMENTED** [review]
> Reviewed 1 of 8 files at r2, 3 of 9 files at r3, 5 of 5 files at r4, 2 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: all files reviewed, 6 unresolved discussions (waiting on @DamianDuy, @kswiecicki, @lplewa, and @pbalcer)

**Review by DamianDuy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: 11 of 12 files reviewed, 6 unresolved discussions (waiting on @blazej-smorawski, @kswiecicki, @lplewa, and @pbalcer)

___
*[`src/runtime.c` line 119 at r4](https://reviewable.io/reviews/pmem/miniasync/100#-N7LT_zZ1PXJfIro3pwS:-N7Q8Z3cCufawKX1YMcV:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/5daae88e9d95e68a77def58acb0fec8e4d5445b0/src/runtime.c#L119)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

use quick sort from libc (sorry, my fault - I should have been more clear) ;)
https://man7.org/linux/man-pages/man3/qsort.3.html 
</blockquote></details>

Done.

**Review by kswiecicki — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r1, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: all files reviewed, 8 unresolved discussions (waiting on @DamianDuy, @lplewa, and @pbalcer)

___
*[`src/include/libminiasync/future.h` line 385 at r7](https://reviewable.io/reviews/pmem/miniasync/100#-N7V6b8I2dB9Ps6E1sJ8:-N7V6b8I2dB9Ps6E1sJ9:b9zz676) ([raw file](https://github.com/pmem/miniasync/blob/9895fe9189d2f0957cb8391c6471183bdfbedaee/src/include/libminiasync/future.h#L385)):*
> ```objc
> 
> static inline int future_has_property_default(void *future,
> 		enum FUTURE_PROPERTY property)
> ```

A new line precedes every function name in miniasync, it would be better if our code were uniform throughout the library.



_Code quote:_
```objc
static inline int future_has_property_default(void *future,
		enum FUTURE_PROPERTY property)
```
___
*[`src/include/libminiasync/future.h` line 392 at r7](https://reviewable.io/reviews/pmem/miniasync/100#-N7V6cM79UT5icwYq3Sf:-N7V6cM79UT5icwYq3Sg:b3kqx) ([raw file](https://github.com/pmem/miniasync/blob/9895fe9189d2f0957cb8391c6471183bdfbedaee/src/include/libminiasync/future.h#L392)):*
> ```objc
> 
> static inline int future_chain_has_property(void *future,
> 		enum FUTURE_PROPERTY property)
> ```

.



_Code quote:_
```objc
static inline int future_chain_has_property(void *future,
		enum FUTURE_PROPERTY property)
```

**Review by DamianDuy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: 11 of 12 files reviewed, 8 unresolved discussions (waiting on @blazej-smorawski, @DamianDuy, @kswiecicki, @lplewa, and @pbalcer)

___
*[`src/include/libminiasync/future.h` line 385 at r7](https://reviewable.io/reviews/pmem/miniasync/100#-N7V6b8I2dB9Ps6E1sJ8:-N7V9zsT4Dc9VAiGqcpK:bbnp06w) ([raw file](https://github.com/pmem/miniasync/blob/9895fe9189d2f0957cb8391c6471183bdfbedaee/src/include/libminiasync/future.h#L385)):*
<details><summary><i>Previously, kswiecicki (Krzysztof Święcicki) wrote…</i></summary><blockquote>

A new line precedes every function name in miniasync, it would be better if our code were uniform throughout the library.
</blockquote></details>

Done. I'm surprised cstyle didn't report it.

**Review by kswiecicki — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: all files reviewed, 6 unresolved discussions (waiting on @lplewa and @pbalcer)

**Review by pbalcer — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100#-:-N7tucZj6GaEuV1gzxxj:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @DamianDuy and @lplewa)

___
*[`src/include/libminiasync/future.h` line 118 at r8](https://reviewable.io/reviews/pmem/miniasync/100#-N7tuQVDEdurhzWcUHJw:-N7tuQVDEdurhzWcUHJx:bo2sbdt) ([raw file](https://github.com/pmem/miniasync/blob/55143d89b2f31eed0573ff78f46748c3173dd3c2/src/include/libminiasync/future.h#L118)):*
> ```objc
> };
> 
> enum FUTURE_PROPERTY {
> ```

use lower case `enum future_property`
___
*[`src/include/libminiasync/future.h` line 337 at r8](https://reviewable.io/reviews/pmem/miniasync/100#-N7tuCtfAF12JkCpzNqw:-N7tuCtfAF12JkCpzNqx:bax2wlc) ([raw file](https://github.com/pmem/miniasync/blob/55143d89b2f31eed0573ff78f46748c3173dd3c2/src/include/libminiasync/future.h#L337)):*
> ```objc
> async_chain_impl(struct future_context *ctx, struct future_notifier *notifier)
> {
> #define _MINIASYNC_PTRSIZE sizeof(void *)
> ```

these macros are no longer needed here

**Review by DamianDuy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/miniasync/100)* status: 4 of 12 files reviewed, 4 unresolved discussions (waiting on @blazej-smorawski, @kswiecicki, @lplewa, and @pbalcer)

___
*[`src/include/libminiasync/future.h` line 118 at r8](https://reviewable.io/reviews/pmem/miniasync/100#-N7tuQVDEdurhzWcUHJw:-N7tzHFWAHBNre4Fv88q:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/55143d89b2f31eed0573ff78f46748c3173dd3c2/src/include/libminiasync/future.h#L118)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

use lower case `enum future_property`
</blockquote></details>

Done.
___
*[`src/include/libminiasync/future.h` line 337 at r8](https://reviewable.io/reviews/pmem/miniasync/100#-N7tuCtfAF12JkCpzNqw:-N7tzJzK7wjM823D7p81:b-896fix) ([raw file](https://github.com/pmem/miniasync/blob/55143d89b2f31eed0573ff78f46748c3173dd3c2/src/include/libminiasync/future.h#L337)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

these macros are no longer needed here
</blockquote></details>

Done.

**Pull request merged**